equational_theories/EquationalResult.lean

```diff
@@ -0,0 +1,103 @@
+  | unconditional : String → EntryVariant
+deriving Lean.ToJson, Lean.FromJson
+
+/-- An entry in the equational results environment extension -/
+structure Entry where
+/-- Name of the declaration. -/
+(name : Name)
+/-- Name of the file where this declaration was found. -/
+(filename : String)
+/-- Which kind of result is it? -/
+(variant : EntryVariant)
+deriving Lean.ToJson, Lean.FromJson
+
+initialize equationalResultsExtension : SimplePersistentEnvExtension Entry (Array Entry) ←
+  registerSimplePersistentEnvExtension {
+    name := `equational_result
+    addImportedFn := Array.concatMap id
+    addEntryFn    := Array.push
+  }
+
+private def equationalResultHelpString : String :=
+"tags theorems that provide implications or negated implications to the directed graph"
+
+/-- Initialization of `equational_result` attribute -/
+initialize equationalResultAttr : Unit ←
+  registerBuiltinAttribute {
+    name  := `equational_result
+    descr := equationalResultHelpString
+    applicationTime := AttributeApplicationTime.afterCompilation
+    add   := fun declName _stx _attrKind => do
+       let filename := (←read).fileName
+       discard <| Meta.MetaM.run do
+       let info ← getConstInfo declName
+       let entry ← match info with
+                   | .thmInfo  (val : TheoremVal) =>
+                     if let some imp ← parseImplication val.type then
+                       pure <| ⟨val.name, filename, .implication imp⟩
+                     else if let some nimp ← parseNonimplication val.type then
+                       pure <| ⟨val.name, filename, .nonimplication nimp⟩
+                     else if let some uncond ← parseUnconditionalEquation val.type then
+                       pure <| ⟨val.name, filename, .unconditional uncond⟩
+                     else
+                       throwError "failed to match type of @[equational_result] theorem"
+                   | _ => throwError "@[equational_result] is only allowed on theorems"
+       modifyEnv fun env =>
+         equationalResultsExtension.addEntry env entry
+    erase := fun _declName =>
+      throwError "can't remove `equational_result` attribute (not implemented yet)"
+  }
+
+/-- Returns the contents of the equational results environment extension.
+-/
+def extractEquationalResults {m : Type → Type} [Monad m] [MonadEnv m] [MonadError m] :
+    m (Array Entry) := do
+  return equationalResultsExtension.getState (← getEnv)
+
+/-- Prints the contents of the equational results environment extension.
+-/
+syntax (name := printEquationalResults) "#print_equational_results" : command
+
+elab_rules : command
+| `(command| #print_equational_results) => do
+  let rs ← extractEquationalResults
+  for ⟨name, _filename, res⟩ in rs do
+    match res with
+    | .implication ⟨lhs, rhs⟩ => println! "{name}: {lhs} → {rhs}"
+    | .nonimplication ⟨lhs, rhs⟩ => println! "{name}: ¬ ({lhs} → {rhs})"
+    | .unconditional rhs => println! "{name}: {rhs} holds unconitionally"
```

Comment:
```suggestion
    | .unconditional rhs => println! "{name}: {rhs} holds unconditionally"
```

Reply:
done in https://github.com/teorth/equational_theories/pull/93/commits/a4d113eef09a2db84f5bf19aee4fceced68e2d17